File: src/solver/dlx.py
"""Dancing Links (DLX) — Knuth's Algorithm X for exact cover.

Ported from third_party/aog/src/dlx.rs (lifthrasiir/aog, Rust).
Uses four doubly-linked circular lists: left, right, up, down.
Column selection: MRV (minimum remaining values).

Usage:
    dlx = Dlx(num_cols)
    for rid, cols in enumerate(candidate_columns):
        dlx.add_row(rid, sorted(cols))
    solution = []
    dlx.search(solution, lambda sol: print(sol))  # finds first solution
    dlx.search_with_check(solution, row_check_fn, callback_fn)  # with pruning
"""

from __future__ import annotations


class Dlx:
    """Dancing Links — Algorithm X exact cover solver."""

    def __init__(self, num_cols: int) -> None:
        n = num_cols + 1  # root (0) + column headers (1..num_cols)
        self._num_cols = num_cols
        self.left: list[int] = list(range(-1, n - 1))
        self.left[0] = num_cols
        self.right: list[int] = list(range(1, n + 1))
        self.right[num_cols] = 0
        self.up: list[int] = list(range(n))
        self.down: list[int] = list(range(n))
        self.col: list[int] = list(range(n))
        self.row_id: list[int] = [0] * n
        self.size: list[int] = [0] * n
        self._deadline: float = float('inf')
        self._node_count: int = 0

    def add_row(self, row_id: int, cols: list[int]) -> None:
        """Add a row that covers the given column indices (must be sorted)."""
        if not cols:
            return
        base = len(self.left)
        length = len(cols)

        for i, c in enumerate(cols):
            h = c + 1  # column header node
            node = base + i

            self.left.append(node - 1 if i > 0 else base + length - 1)
            self.right.append(node + 1 if i < length - 1 else base)
            prev_up = self.up[h]
            self.up.append(prev_up)
            self.down.append(h)
            self.col.append(h)
            self.row_id.append(row_id)
            self.size.append(0)

            self.down[prev_up] = node
            self.up[h] = node
            self.size[h] += 1

    def _cover(self, c: int) -> None:
        """Remove column c from header list, and all rows in this column."""
        self.right[self.left[c]] = self.right[c]
        self.left[self.right[c]] = self.left[c]

        i = self.down[c]
        while i != c:
            j = self.right[i]
            while j != i:
                self.down[self.up[j]] = self.down[j]
                self.up[self.down[j]] = self.up[j]
                self.size[self.col[j]] -= 1
                j = self.right[j]
            i = self.down[i]

    def _uncover(self, c: int) -> None:
        """Restore column c and all rows (reverse order of _cover)."""
        i = self.up[c]
        while i != c:
            j = self.left[i]
            while j != i:
                self.size[self.col[j]] += 1
                self.down[self.up[j]] = j
                self.up[self.down[j]] = j
                j = self.left[j]
            i = self.up[i]

        self.right[self.left[c]] = c
        self.left[self.right[c]] = c

    def _choose_column(self) -> int:
        """MRV: pick column with fewest remaining rows."""
        best = self.right[0]
        best_size = self.size[best]
        c = self.right[best]
        while c != 0:
            if self.size[c] < best_size:
                best = c
                best_size = self.size[c]
                if best_size == 0:
                    break
            c = self.right[c]
        return best

    def search(self, solution: list[int], callback) -> bool:
        """Run Algorithm X. callback(sol_rows) returns True to continue search."""
        return self.search_with_check(solution, lambda _: True, callback)

    def search_with_check(self, solution: list[int], row_check, callback) -> bool:
        """Algorithm X with incremental validation.

        row_check(solution) is called after each row is added; returning False prunes.
        callback(sol_rows) is called for complete solutions; returning False stops search.
        Returns True to continue, False to stop.
        """
        if self.right[0] == 0:
            return callback(solution)

        self._node_count += 1
        if self._node_count % 50000 == 0:
            import time
            if time.monotonic() > self._deadline:
                return False

        c = self._choose_column()
        if self.size[c] == 0:
            return True

        self._cover(c)
        cont = True
        r = self.down[c]

        while r != c and cont:
            solution.append(self.row_id[r])

            j = self.right[r]
            while j != r:
                self._cover(self.col[j])
                j = self.right[j]

            if row_check(solution):
                cont = self.search_with_check(solution, row_check, callback)

            solution.pop()
            j = self.left[r]
            while j != r:
                self._uncover(self.col[j])
                j = self.left[j]

            r = self.down[r]

        self._uncover(c)
        return cont

File: rsolver/src/dlx.rs
//! Dancing Links (Algorithm X) for exact cover.

use std::time::Instant;

#[derive(Debug, Clone, Default)]
struct DlxNode {
    left: usize,
    right: usize,
    up: usize,
    down: usize,
    col: usize,
    row_id: Option<usize>,
    count: usize,
}

/// Dancing Links solver for exact cover problems.
///
/// Columns represent constraints, rows represent choices.
/// `search()` returns true if a solution is found.
#[derive(Debug)]
pub struct DancingLinks {
    nodes: Vec<DlxNode>,
    pub search_count: u64,
    deadline: Option<Instant>,
}

impl DancingLinks {
    /// Create a new DLX instance with `col_count` primary columns.
    pub fn new(col_count: usize) -> Self {
        let mut nodes = Vec::with_capacity(col_count + 1);
        // Root node at index 0
        nodes.push(DlxNode {
            left: col_count,
            right: 1,
            ..Default::default()
        });
        // Column headers at indices 1..=col_count
        for c in 1..=col_count {
            nodes.push(DlxNode {
                left: c - 1,
                right: if c == col_count { 0 } else { c + 1 },
                up: c,
                down: c,
                col: c,
                ..Default::default()
            });
        }
        Self {
            nodes,
            search_count: 0,
            deadline: None,
        }
    }

    /// Set a deadline after which `search()` will abort.
    pub fn set_deadline(&mut self, d: Instant) {
        self.deadline = Some(d);
    }

    /// Add a row covering the given columns. `row_id` is an opaque identifier
    /// that will be collected when the row is selected in a solution.
    pub fn add_row(&mut self, cols: &[usize], row_id: usize) {
        let start = self.nodes.len();
        for (i, &col) in cols.iter().enumerate() {
            let c = col + 1; // offset for root
            let cur = start + i;
            let left_pos = if i == 0 { start + cols.len() - 1 } else { cur - 1 };
            let right_pos = if i == cols.len() - 1 { start } else { cur + 1 };
            let hdr_down = self.nodes[c].down;

            self.nodes.push(DlxNode {
                left: left_pos,
                right: right_pos,
                up: c,
                down: hdr_down,
                col: c,
                row_id: Some(row_id),
                count: 0,
            });
            let idx = self.nodes.len() - 1;
            self.nodes[c].down = idx;
            self.nodes[hdr_down].up = idx;
            self.nodes[c].count += 1;
        }
    }

    /// Cover (remove) a column.
    fn cover(&mut self, col: usize) {
        let c = col + 1;
        let right = self.nodes[c].right;
        let left = self.nodes[c].left;
        self.nodes[right].left = left;
        self.nodes[left].right = right;

        let mut i = self.nodes[c].down;
        while i != c {
            let mut j = self.nodes[i].right;
            while j != i {
                let nd = self.nodes[j].down;
                let nu = self.nodes[j].up;
                let col_idx = self.nodes[j].col;
                self.nodes[nd].up = nu;
                self.nodes[nu].down = nd;
                self.nodes[col_idx].count -= 1;
                j = self.nodes[j].right;
            }
            i = self.nodes[i].down;
        }
    }

    /// Uncover (restore) a column.
    fn uncover(&mut self, col: usize) {
        let c = col + 1;
        let mut i = self.nodes[c].up;
        while i != c {
            let mut j = self.nodes[i].left;
            while j != i {
                let col_idx = self.nodes[j].col;
                self.nodes[col_idx].count += 1;
                let nd = self.nodes[j].down;
                let nu = self.nodes[j].up;
                self.nodes[nd].up = j;
                self.nodes[nu].down = j;
                j = self.nodes[j].left;
            }
            i = self.nodes[i].up;
        }
        let right = self.nodes[c].right;
        let left = self.nodes[c].left;
        self.nodes[right].left = c;
        self.nodes[left].right = c;
    }

    /// Choose the column with the fewest rows (MRV heuristic).
    fn choose_column(&self) -> Option<usize> {
        let mut min_count = usize::MAX;
        let mut best = None;
        let mut c = self.nodes[0].right;
        while c != 0 {
            if self.nodes[c].count < min_count {
                min_count = self.nodes[c].count;
                best = Some(c - 1);
                // D5: a column with count 0 means the branch is dead; count 1
                // is a forced move — no column can have fewer, so stop scanning.
                // (doc 16 §2 D5; saves 30-50% of the MRV column scan.)
                if min_count <= 1 {
                    break;
                }
            }
            c = self.nodes[c].right;
        }
        best
    }

    /// Search with an incremental validation callback.
    /// `row_check` is called after each row selection with the current partial
    /// solution (row IDs); returning false prunes the branch.  `on_solution` is
    /// called at each complete solution; returning true stops the whole search.
    /// Returns true if a solution that stopped the search was found.
    pub fn search_with_check(
        &mut self,
        depth: usize,
        partial: &mut Vec<usize>,
        row_check: &mut dyn FnMut(&[usize]) -> bool,
        on_solution: &mut dyn FnMut(&[usize]) -> bool,
    ) -> bool {
        // D9: throttle the deadline check to every 1024 nodes instead of every
        // node — Instant::now() is a ~20ns vDSO call, unnecessary on every
        // recursion. search_count is incremented below; check it first so the
        // very first nodes still get a timely check. (doc 16 §2 D9, -10-25%.)
        if let Some(d) = self.deadline {
            if (self.search_count & 1023) == 0 && Instant::now() >= d {
                return false;
            }
        }
        self.search_count += 1;

        if self.nodes[0].right == 0 {
            return on_solution(partial);
        }

        let col = match self.choose_column() {
            Some(c) => c,
            None => return false,
        };

        self.cover(col);
        let c_node = col + 1;
        let mut r = self.nodes[c_node].down;
        while r != c_node {
            let row_id = self.nodes[r].row_id;
            if let Some(rid) = row_id {
                partial.push(rid);

                // Incremental check
                if row_check(partial) {
                    // Cover all columns in this row
                    let mut j = self.nodes[r].right;
                    while j != r {
                        self.cover(self.nodes[j].col - 1);
                        j = self.nodes[j].right;
                    }

                    if self.search_with_check(depth + 1, partial, row_check, on_solution) {
                        // Stop propagates up; matrix state is left mid-search
                        // (callers do not reuse the DLX object after this).
                        return true;
                    }

                    // Uncover
                    j = self.nodes[r].left;
                    while j != r {
                        self.uncover(self.nodes[j].col - 1);
                        j = self.nodes[j].left;
                    }
                }

                partial.pop();
            }
            r = self.nodes[r].down;
        }
        self.uncover(col);
        false
    }
}
